### ISSUE TITLE
Global values defined in a subchart can affect global values in parent chart and sibling charts

### ISSUE THREAD

**Issue by bgutjahr** [helm/helm#9332]
The issue can be reproduced with a chart that contains 2 subcharts and the following values:

1. toplevel values:
    ```
    global:
      nested:
        v1: chart
    ```

2. values in subchart alpha:
    ```
    global:
      nested:
        v2: alpha
        v3: alpha
    ```
3. values in subchart beta:
    ```
    global:
      nested:
        v2: beta
        v4: beta
    ```

After installing such a chart, `helm get values -a` produced the following output:

```
COMPUTED VALUES:
alpha:
  global:
    nested:
      v1: chart
      v2: alpha
      v3: alpha
      v4: beta
beta:
  global:
    nested:
      v1: chart
      v2: alpha
      v3: alpha
      v4: beta
global:
  nested:
    v1: chart
    v2: alpha
    v3: alpha
    v4: beta

```

This shows that values from the subcharts become visible in the toplevel chart and in other subcharts,

In my opinion, this behavior contradicts the following statement, which can be found in https://helm.sh/docs/topics/charts/:
> "There is no way for a subchart to influence the values of the parent chart."

Here is also a zip file that contains my example chart to reproduce the issue: [chart.zip](https://github.com/helm/helm/files/5933020/chart.zip)

Output of `helm version`:
```
version.BuildInfo{Version:"v3.3.0", GitCommit:"8a4aeec08d67a7b84472007529e8097ec3742105", GitTreeState:"dirty", GoVersion:"go1.14.7"}
```

Output of `kubectl version`:
```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:30:33Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:23:04Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
```

Cloud Provider/Platform (AKS, GKE, Minikube etc.): N/A

**Comment by technosophos** [comment]
I have marked this as a bug... but I am not entirely sure whether it is.

The docs are pretty unequivocal about this:

> If a subchart declares a global variable, that global will be passed downward (to the subchart's subcharts), but not upward to the parent chart. There is no way for a subchart to influence the values of the parent chart.
> https://helm.sh/docs/topics/charts/#global-values

However, those docs are also quite old (2.0.0) and could be incorrect. Does anyone know if that behavior was intentionally changed in Helm 3, or if this is a mistake?

**Comment by bacongobbler** [comment]
I do not recall there being any intentional changes to the global values' behavior for Helm 3, so I think you're right in assuming this is a bug.

**Comment by bgutjahr** [comment]
Hello, together with the submittal of this issue, I had also create a PR that fixes it. So far, I haven't seen any progress on the PR (like assigned reviewers or review comments). So I wonder if I had missed to provided some more details that would be necessary to get  the PR to be considered.

**Comment by bacongobbler** [comment]
Nope, you haven't missed anything. The project is run by volunteers and community members sponsored by their employers. PRs take time to review.

Thank you for being patient.

**Comment by bgutjahr** [comment]
Instead of autoclosing the PR, I wonder if someone could eventually have a look at my changes and either give feedback or just merge it. The actual fix is pretty short, all the rest are new unit tests, plus additional improvements to prevent similar issues in that code.
I could also replace my PR with just the actual one-line-fix (without unit tests and without the improvements), if that would get it reviewed and merged faster.

**Comment by technosophos** [comment]
The issue was that the core maintainers never reached consensus on whether the change was intentional. If it was, then the docs need to be changed. If it was not, then we need to make sure this does not run across our breaking change guidelines, then we can merge it.

If you have a moment, @adamreese , can you check this and see if the bug reported here is a regression, or was introduced on purpose?

**Comment by bgutjahr** [comment]
Please have a look at my PR's 2nd commit (4fafe2fe114ea25331e2836a6c146f04286c38e9), which contains the actual fix. In my opinion, the fix shows that this doesn't seem to have been implemented on purpose.

**Comment by bgutjahr** [comment]
I have to repeat my comment for end of May:

_Instead of autoclosing the PR, I wonder if someone could eventually have a look at my changes and either give feedback or just merge it.
The actual fix is pretty short, all the rest are new unit tests, plus additional improvements to prevent similar issues in that code._

It this behavoir was considered a feature, I would really like to see how you want to have that behavior documented. And I really doubt it, since my fix in commit 4fafe2f shows that the cause was just a missing continue statement, which renderer the codelines(s) above ineffective.

**Comment by hickeyma** [comment]
Realted #9749
